### PR TITLE
gh-111354: Simplify _PyGen_yf by moving some of its work to the compiler and frame state

### PR DESCRIPTION
**Pull request by iritkatriel**
This splits the SUSPENDED frame state into SUSPENDED and SUSPENDED_YIELD_FROM, adds an oparg to YIELD_VALUE to indicate whether it's in a yield-from or await (so it can set the state correctly). With this, `_PyGen_yf` has a lot less work to do.

* Issue: gh-111354

----
:books: Documentation preview :books:: https://cpython-previews--111648.org.readthedocs.build/

### CODE REVIEW

File: Python/bytecodes.c [hunk 1]
```diff
@@ -1049,7 +1049,7 @@ dummy_func(
             assert(frame != &entry_frame);
             frame->instr_ptr = next_instr;
             PyGenObject *gen = _PyFrame_GetGenerator(frame);
-            gen->gi_frame_state = FRAME_SUSPENDED;
+            gen->gi_frame_state = oparg ? FRAME_SUSPENDED_YIELD_FROM : FRAME_SUSPENDED;
```
Comment (markshannon) [inline Python/bytecodes.c]: Can we make this unconditional?
Ideally `gen->gi_frame_state = oparg`, but that might require flipping the sign of the all the gen state constants.
`gen->gi_frame_state = oparg + K` would be OK.

File: Python/bytecodes.c [hunk 2]
```diff
@@ -1075,7 +1075,7 @@ dummy_func(
...
```
Comment (markshannon) [inline Python/bytecodes.c]: Ditto

**Review by markshannon — CHANGES_REQUESTED** [review]
I'd like to avoid the extra branch in `YIELD_VALUE`.
Other than that, looks good.

**Comment by iritkatriel** [comment]
When you're done making the requested changes, leave the comment: I have made the requested changes; please review again.

**Review by markshannon — APPROVED** [review]
Looks good.

**Pull request merged**